File: scripts/server/team_server/src/team_manager.cpp
#include "team_manager.h"
#include "message_ids.h"
#include "log_macros.h"

#include <chrono>
#include <algorithm>

namespace farm {

TeamManager::TeamManager(SendToPlayerFunc send_func)
    : send_func_(std::move(send_func))
{
}

bool TeamManager::init(const std::string& redis_host, int redis_port) {
    return redis_.connect(redis_host, redis_port);
}

int TeamManager::create_team(uint64_t player_id) {
    // Check if already in team
    if (is_in_team(player_id)) {
        return static_cast<int>(farm::TEAM_ALREADY_IN_TEAM);
    }

    uint64_t team_id = redis_.generate_team_id();
    if (team_id == 0) {
        SPDLOG_ERROR("[Team]Failed to generate team_id");
        return -1;
    }

    // Create in Redis
    auto now = std::chrono::duration_cast<std::chrono::seconds>(
        std::chrono::system_clock::now().time_since_epoch()).count();

    redis_.set_team_info(team_id, player_id, 0, 0);
    redis_.add_team_member(team_id, player_id);
    redis_.set_player_team(player_id, team_id);

    // Create in memory
    TeamInfo team;
    team.team_id = team_id;
    team.leader_id = player_id;
    team.members.push_back(player_id);
    team.status = TeamStatus::IDLE;
    team.cave_level = 0;
    team.created_at = now;
    teams_[team_id] = team;
    player_team_map_[player_id] = team_id;

    SPDLOG_INFO("[Team]Player {} created team {}", player_id, team_id);

    // Send response
    farm::TeamCreateResp resp;
    resp.set_code(farm::TEAM_SUCCESS);
    resp.set_team_id(team_id);
    std::string data;
    resp.SerializeToString(&data);
    send_func_(player_id, MSG_ID_TEAM_CREATE_RESP,
               reinterpret_cast<const uint8_t*>(data.data()), data.size());

    return static_cast<int>(farm::TEAM_SUCCESS);
}

int TeamManager::disband_team(uint64_t player_id) {
    auto it = player_team_map_.find(player_id);
    if (it == player_team_map_.end()) {
        return static_cast<int>(farm::TEAM_NOT_IN_TEAM);
    }

    uint64_t team_id = it->second;
    auto team_it = teams_.find(team_id);
    if (team_it == teams_.end()) {
        return static_cast<int>(farm::TEAM_NOT_IN_TEAM);
    }

    TeamInfo& team = team_it->second;
    if (team.leader_id != player_id) {
        return static_cast<int>(farm::TEAM_NOT_LEADER);
    }
    if (team.status == TeamStatus::IN_CAVE) {
        return static_cast<int>(farm::TEAM_IN_CAVE);
    }

    // Notify all members
    farm::TeamDisbandResp resp;
    resp.set_code(farm::TEAM_SUCCESS);
    std::string data;
    resp.SerializeToString(&data);

    for (uint64_t mid : team.members) {
        send_func_(mid, MSG_ID_TEAM_DISBAND_RESP,
                   reinterpret_cast<const uint8_t*>(data.data()), data.size());
        player_team_map_.erase(mid);
        redis_.del_player_team(mid);
    }

    // Clean up Redis
    for (uint64_t mid : team.members) {
        redis_.remove_team_member(team_id, mid);
    }
    redis_.del_team_info(team_id);

    // Remove from memory
    teams_.erase(team_it);

    SPDLOG_INFO("[Team]Team {} disbanded by player {}", team_id, player_id);
    return static_cast<int>(farm::TEAM_SUCCESS);
}

int TeamManager::invite_player(uint64_t inviter_id, uint64_t target_id) {
    if (inviter_id == target_id) {
        return static_cast<int>(farm::TEAM_SELF_OPERATION);
    }

    auto it = player_team_map_.find(inviter_id);
    if (it == player_team_map_.end()) {
        return static_cast<int>(farm::TEAM_NOT_IN_TEAM);
    }

    uint64_t team_id = it->second;
    auto team_it = teams_.find(team_id);
    if (team_it == teams_.end()) {
        return static_cast<int>(farm::TEAM_NOT_IN_TEAM);
    }

    TeamInfo& team = team_it->second;
    if (team.leader_id != inviter_id) {
        return static_cast<int>(farm::TEAM_NOT_LEADER);
    }
    if (team.members.size() >= MAX_TEAM_SIZE) {
        return static_cast<int>(farm::TEAM_FULL);
    }
    if (is_in_team(target_id)) {
        return static_cast<int>(farm::TEAM_TARGET_ALREADY_IN_TEAM);
    }

    // Store invite in Redis
    auto now = std::chrono::duration_cast<std::chrono::seconds>(
        std::chrono::system_clock::now().time_since_epoch()).count();
    redis_.set_invite(target_id, team_id, inviter_id, now);

    SPDLOG_INFO("[Team]Player {} invited player {} to team {}",
                inviter_id, target_id, team_id);

    // Send invite notification to target
    farm::TeamInviteNotify notify;
    notify.set_team_id(team_id);
    notify.set_inviter_id(inviter_id);
    // TODO: set inviter_name and inviter_level from player data
    std::string notify_data;
    notify.SerializeToString(&notify_data);
    send_func_(target_id, MSG_ID_TEAM_INVITE_NOTIFY,
               reinterpret_cast<const uint8_t*>(notify_data.data()), notify_data.size());

    // Send response to inviter
    farm::TeamInviteResp resp;
    resp.set_code(farm::TEAM_SUCCESS);
    std::string resp_data;
    resp.SerializeToString(&resp_data);
    send_func_(inviter_id, MSG_ID_TEAM_INVITE_RESP,
               reinterpret_cast<const uint8_t*>(resp_data.data()), resp_data.size());

    return static_cast<int>(farm::TEAM_SUCCESS);
}

int TeamManager::accept_invite(uint64_t player_id, uint64_t team_id) {
    // Check invite exists
    uint64_t inv_team_id = 0, inviter_id = 0, timestamp = 0;
    if (!redis_.get_invite(player_id, inv_team_id, inviter_id, timestamp)) {
        return static_cast<int>(farm::TEAM_INVITE_NOT_FOUND);
    }
    if (inv_team_id != team_id) {
        return static_cast<int>(farm::TEAM_INVITE_NOT_FOUND);
    }

    // Check if already in team (race condition)
    if (is_in_team(player_id)) {
        redis_.del_invite(player_id);
        return static_cast<int>(farm::TEAM_ALREADY_IN_TEAM);
    }

    auto team_it = teams_.find(team_id);
    if (team_it == teams_.end()) {
        redis_.del_invite(player_id);
        return static_cast<int>(farm::TEAM_INVITE_NOT_FOUND);
    }

    TeamInfo& team = team_it->second;
    if (team.members.size() >= MAX_TEAM_SIZE) {
        redis_.del_invite(player_id);
        return static_cast<int>(farm::TEAM_FULL);
    }

    // Add to team
    team.members.push_back(player_id);
    player_team_map_[player_id] = team_id;
    redis_.add_team_member(team_id, player_id);
    redis_.set_player_team(player_id, team_id);
    redis_.del_invite(player_id);

    SPDLOG_INFO("[Team]Player {} joined team {}", player_id, team_id);

    // Notify all members
    notify_member_update(team_id, 0 /* JOIN */, player_id, "", 0);

    // Send accept response with full member list
    farm::TeamAcceptResp resp;
    resp.set_code(farm::TEAM_SUCCESS);
    for (uint64_t mid : team.members) {
        auto* member = resp.add_members();
        member->set_player_id(mid);
        member->set_is_leader(mid == team.leader_id);
        // TODO: populate role_name, level, online, hp
    }
    std::string resp_data;
    resp.SerializeToString(&resp_data);
    send_func_(player_id, MSG_ID_TEAM_ACCEPT_RESP,
               reinterpret_cast<const uint8_t*>(resp_data.data()), resp_data.size());

    return static_cast<int>(farm::TEAM_SUCCESS);
}

int TeamManager::reject_invite(uint64_t player_id, uint64_t team_id) {
    redis_.del_invite(player_id);

    farm::TeamRejectResp resp;
    resp.set_code(farm::TEAM_SUCCESS);
    std::string data;
    resp.SerializeToString(&data);
    send_func_(player_id, MSG_ID_TEAM_REJECT_RESP,
               reinterpret_cast<const uint8_t*>(data.data()), data.size());

    return static_cast<int>(farm::TEAM_SUCCESS);
}

int TeamManager::leave_team(uint64_t player_id) {
    auto it = player_team_map_.find(player_id);
    if (it == player_team_map_.end()) {
        return static_cast<int>(farm::TEAM_NOT_IN_TEAM);
    }

    uint64_t team_id = it->second;
    auto team_it = teams_.find(team_id);
    if (team_it == teams_.end()) {
        player_team_map_.erase(it);
        return static_cast<int>(farm::TEAM_NOT_IN_TEAM);
    }

    TeamInfo& team = team_it->second;

    // Remove from team
    team.members.erase(
        std::remove(team.members.begin(), team.members.end(), player_id),
        team.members.end());
    player_team_map_.erase(player_id);
    redis_.remove_team_member(team_id, player_id);
    redis_.del_player_team(player_id);

    // If leader left, transfer to next member
    if (team.leader_id == player_id && !team.members.empty()) {
        uint64_t new_leader = team.members[0];  // earliest joiner
        team.leader_id = new_leader;
        redis_.set_team_info(team_id, new_leader,
                             static_cast<uint32_t>(team.status), team.cave_level);

        SPDLOG_INFO("[Team]Leader transferred to player {} in team {}",
                    new_leader, team_id);

        // Notify leader change
        notify_leader_change(team_id, new_leader, "");
    }

    // If team is empty, disband
    if (team.members.empty()) {
        redis_.del_team_info(team_id);
        teams_.erase(team_it);
        SPDLOG_INFO("[Team]Team {} auto-disbanded (empty)", team_id);
    } else {
        // Notify remaining members
        notify_member_update(team_id, 1 /* LEAVE */, player_id, "", 0);
    }

    // Send response to leaving player
    farm::TeamLeaveResp resp;
    resp.set_code(farm::TEAM_SUCCESS);
    std::string data;
    resp.SerializeToString(&data);
    send_func_(player_id, MSG_ID_TEAM_LEAVE_RESP,
               reinterpret_cast<const uint8_t*>(data.data()), data.size());

    return static_cast<int>(farm::TEAM_SUCCESS);
}

int TeamManager::kick_member(uint64_t leader_id, uint64_t target_id) {
    if (leader_id == target_id) {
        return static_cast<int>(farm::TEAM_CANNOT_KICK_SELF);
    }

    auto it = player_team_map_.find(leader_id);
    if (it == player_team_map_.end()) {
        return static_cast<int>(farm::TEAM_NOT_IN_TEAM);
    }

    uint64_t team_id = it->second;
    auto team_it = teams_.find(team_id);
    if (team_it == teams_.end()) {
        return static_cast<int>(farm::TEAM_NOT_IN_TEAM);
    }

    TeamInfo& team = team_it->second;
    if (team.leader_id != leader_id) {
        return static_cast<int>(farm::TEAM_NOT_LEADER);
    }

    // Check target is in team
    auto target_it = std::find(team.members.begin(), team.members.end(), target_id);
    if (target_it == team.members.end()) {
        return static_cast<int>(farm::TEAM_TARGET_NOT_FOUND);
    }

    // Remove target
    team.members.erase(target_it);
    player_team_map_.erase(target_id);
    redis_.remove_team_member(team_id, target_id);
    redis_.del_player_team(target_id);

    SPDLOG_INFO("[Team]Player {} kicked player {} from team {}",
                leader_id, target_id, team_id);

    // Notify kicked player
    farm::TeamMemberUpdate kick_notify;
    kick_notify.set_update_type(farm::TEAM_UPDATE_KICK);
    auto* member = kick_notify.mutable_member();
    member->set_player_id(target_id);
    std::string kick_data;
    kick_notify.SerializeToString(&kick_data);
    send_func_(target_id, MSG_ID_TEAM_MEMBER_UPDATE,
               reinterpret_cast<const uint8_t*>(kick_data.data()), kick_data.size());

    // Notify remaining members
    notify_member_update(team_id, 2 /* KICK */, target_id, "", 0);

    // Send response to leader
    farm::TeamKickResp resp;
    resp.set_code(farm::TEAM_SUCCESS);
    std::string resp_data;
    resp.SerializeToString(&resp_data);
    send_func_(leader_id, MSG_ID_TEAM_KICK_RESP,
               reinterpret_cast<const uint8_t*>(resp_data.data()), resp_data.size());

    return static_cast<int>(farm::TEAM_SUCCESS);
}

TeamInfo* TeamManager::get_team(uint64_t team_id) {
    auto it = teams_.find(team_id);
    return (it != teams_.end()) ? &it->second : nullptr;
}

TeamInfo* TeamManager::get_player_team(uint64_t player_id) {
    auto it = player_team_map_.find(player_id);
    if (it == player_team_map_.end()) return nullptr;
    return get_team(it->second);
}

bool TeamManager::is_in_team(uint64_t player_id) {
    return player_team_map_.find(player_id) != player_team_map_.end();
}

void TeamManager::check_invite_timeouts() {
    auto now = std::chrono::duration_cast<std::chrono::seconds>(
        std::chrono::system_clock::now().time_since_epoch()).count();

    // This is a simplified check - in production, iterate all pending invites
    // For now, invites are cleaned up on accept/reject/timeout via Redis TTL
}

void TeamManager::notify_member_update(uint64_t team_id, uint32_t update_type,
                                       uint64_t member_id, const std::string& member_name,
                                       uint32_t member_level) {
    auto* team = get_team(team_id);
    if (!team) return;

    farm::TeamMemberUpdate notify;
    notify.set_update_type(static_cast<farm::TeamUpdateType>(update_type));
    auto* member = notify.mutable_member();
    member->set_player_id(member_id);
    member->set_role_name(member_name);
    member->set_level(member_level);

    for (uint64_t mid : team->members) {
        auto* m = notify.add_all_members();
        m->set_player_id(mid);
        m->set_is_leader(mid == team->leader_id);
    }

    std::string data;
    notify.SerializeToString(&data);

    for (uint64_t mid : team->members) {
        send_func_(mid, MSG_ID_TEAM_MEMBER_UPDATE,
                   reinterpret_cast<const uint8_t*>(data.data()), data.size());
    }
}

void TeamManager::notify_leader_change(uint64_t team_id, uint64_t new_leader_id,
                                       const std::string& new_leader_name) {
    auto* team = get_team(team_id);
    if (!team) return;

    farm::TeamLeaderChange notify;
    notify.set_new_leader_id(new_leader_id);
    notify.set_new_leader_name(new_leader_name);

    std::string data;
    notify.SerializeToString(&data);

    for (uint64_t mid : team->members) {
        send_func_(mid, MSG_ID_TEAM_LEADER_CHANGE,
                   reinterpret_cast<const uint8_t*>(data.data()), data.size());
    }
}

void TeamManager::notify_status_update(uint64_t team_id, uint32_t status, int cave_level) {
    auto* team = get_team(team_id);
    if (!team) return;

    farm::TeamStatusUpdate notify;
    notify.set_status(status);
    notify.set_cave_level(cave_level);

    std::string data;
    notify.SerializeToString(&data);

    for (uint64_t mid : team->members) {
        send_func_(mid, MSG_ID_TEAM_STATUS_UPDATE,
                   reinterpret_cast<const uint8_t*>(data.data()), data.size());
    }
}

void TeamManager::load_team_from_redis(uint64_t team_id) {
    // Load team info from Redis into memory
    uint64_t leader_id = 0;
    uint32_t status = 0;
    int cave_level = 0;
    if (!redis_.get_team_info(team_id, leader_id, status, cave_level)) return;

    std::vector<uint64_t> members;
    if (!redis_.get_team_members(team_id, members)) return;

    TeamInfo team;
    team.team_id = team_id;
    team.leader_id = leader_id;
    team.members = members;
    team.status = static_cast<TeamStatus>(status);
    team.cave_level = cave_level;
    teams_[team_id] = team;

    for (uint64_t mid : members) {
        player_team_map_[mid] = team_id;
    }
}

void TeamManager::remove_team_from_memory(uint64_t team_id) {
    auto it = teams_.find(team_id);
    if (it == teams_.end()) return;
    for (uint64_t mid : it->second.members) {
        player_team_map_.erase(mid);
    }
    teams_.erase(it);
}

}  // namespace farm

File: scripts/client/team_manager.py
"""
客户端组队管理器
管理队伍状态、发送组队请求、处理组队回调。
"""
import logging
from typing import List, Optional, Callable
from dataclasses import dataclass, field

from .message_ids import (
    MSG_ID_TEAM_CREATE_REQ, MSG_ID_TEAM_CREATE_RESP,
    MSG_ID_TEAM_DISBAND_REQ, MSG_ID_TEAM_DISBAND_RESP,
    MSG_ID_TEAM_INVITE_REQ, MSG_ID_TEAM_INVITE_RESP,
    MSG_ID_TEAM_INVITE_NOTIFY,
    MSG_ID_TEAM_ACCEPT_REQ, MSG_ID_TEAM_ACCEPT_RESP,
    MSG_ID_TEAM_REJECT_REQ, MSG_ID_TEAM_REJECT_RESP,
    MSG_ID_TEAM_LEAVE_REQ, MSG_ID_TEAM_LEAVE_RESP,
    MSG_ID_TEAM_KICK_REQ, MSG_ID_TEAM_KICK_RESP,
    MSG_ID_TEAM_INFO_REQ, MSG_ID_TEAM_INFO_RESP,
    MSG_ID_TEAM_MEMBER_UPDATE, MSG_ID_TEAM_LEADER_CHANGE,
    MSG_ID_TEAM_STATUS_UPDATE,
)

logger = logging.getLogger("client.team_manager")


@dataclass
class TeamMember:
    """队伍成员信息"""
    player_id: int = 0
    role_name: str = ""
    level: int = 0
    online: bool = True
    is_leader: bool = False
    current_hp: int = 0
    max_hp: int = 0


@dataclass
class TeamInviteNotify:
    """组队邀请通知"""
    team_id: int = 0
    inviter_id: int = 0
    inviter_name: str = ""
    inviter_level: int = 0


class TeamManager:
    """客户端组队管理器"""

    def __init__(self, connection, player_id: int):
        self._connection = connection
        self._player_id = player_id
        self._team_id: int = 0
        self._leader_id: int = 0
        self._members: List[TeamMember] = []
        self._pending_invites: List[TeamInviteNotify] = []
        self._status: int = 0  # 0=idle, 1=in_cave
        self._cave_level: int = 0

        # Callbacks
        self.on_team_created: Optional[Callable[[int], None]] = None
        self.on_team_disbanded: Optional[Callable[[], None]] = None
        self.on_member_update: Optional[Callable[[], None]] = None
        self.on_leader_change: Optional[Callable[[int], None]] = None
        self.on_invite_received: Optional[Callable[[TeamInviteNotify], None]] = None
        self.on_error: Optional[Callable[[int, str], None]] = None

    @property
    def in_team(self) -> bool:
        return self._team_id != 0

    @property
    def is_leader(self) -> bool:
        return self._leader_id == self._player_id

    @property
    def team_id(self) -> int:
        return self._team_id

    @property
    def members(self) -> List[TeamMember]:
        return self._members

    @property
    def member_count(self) -> int:
        return len(self._members)

    @property
    def pending_invites(self) -> List[TeamInviteNotify]:
        return self._pending_invites

    def create_team(self) -> None:
        """创建队伍"""
        logger.info("Creating team")
        self._connection.send_message(MSG_ID_TEAM_CREATE_REQ, b'')

    def disband_team(self) -> None:
        """解散队伍（队长）"""
        if not self.in_team:
            return
        logger.info(f"Disbanding team {self._team_id}")
        self._connection.send_message(MSG_ID_TEAM_DISBAND_REQ, b'')

    def invite_player(self, target_id: int) -> None:
        """邀请玩家组队"""
        if not self.in_team:
            return
        logger.info(f"Inviting player {target_id} to team {self._team_id}")
        # Encode target_id as uint64
        data = target_id.to_bytes(8, 'little')
        self._connection.send_message(MSG_ID_TEAM_INVITE_REQ, data)

    def accept_invite(self, team_id: int) -> None:
        """接受组队邀请"""
        logger.info(f"Accepting invite to team {team_id}")
        data = team_id.to_bytes(8, 'little')
        self._connection.send_message(MSG_ID_TEAM_ACCEPT_REQ, data)

    def reject_invite(self, team_id: int) -> None:
        """拒绝组队邀请"""
        logger.info(f"Rejecting invite to team {team_id}")
        data = team_id.to_bytes(8, 'little')
        self._connection.send_message(MSG_ID_TEAM_REJECT_REQ, data)

    def leave_team(self) -> None:
        """离开队伍"""
        if not self.in_team:
            return
        logger.info(f"Leaving team {self._team_id}")
        self._connection.send_message(MSG_ID_TEAM_LEAVE_REQ, b'')

    def kick_member(self, target_id: int) -> None:
        """踢出成员（队长）"""
        if not self.in_team or not self.is_leader:
            return
        logger.info(f"Kicking player {target_id} from team {self._team_id}")
        data = target_id.to_bytes(8, 'little')
        self._connection.send_message(MSG_ID_TEAM_KICK_REQ, data)

    def request_team_info(self) -> None:
        """查询队伍信息"""
        self._connection.send_message(MSG_ID_TEAM_INFO_REQ, b'')

    # --- Message handlers (called by NetworkDispatcher) ---

    def on_team_create_resp(self, data: bytes) -> None:
        """处理创建队伍响应"""
        import team_pb2
        resp = team_pb2.TeamCreateResp()
        resp.ParseFromString(data)
        if resp.code == 0:
            self._team_id = resp.team_id
            logger.info(f"Team created: {self._team_id}")
            if self.on_team_created:
                self.on_team_created(self._team_id)
        else:
            logger.warning(f"Create team failed: code={resp.code}")
            if self.on_error:
                self.on_error(resp.code, "创建队伍失败")

    def on_team_disband_resp(self, data: bytes) -> None:
        """处理解散队伍响应"""
        import team_pb2
        resp = team_pb2.TeamDisbandResp()
        resp.ParseFromString(data)
        if resp.code == 0:
            self._clear_team()
            logger.info("Team disbanded")
            if self.on_team_disbanded:
                self.on_team_disbanded()

    def on_team_invite_resp(self, data: bytes) -> None:
        """处理邀请响应"""
        import team_pb2
        resp = team_pb2.TeamInviteResp()
        resp.ParseFromString(data)
        if resp.code != 0:
            logger.warning(f"Invite failed: code={resp.code}")
            if self.on_error:
                self.on_error(resp.code, "邀请失败")

    def on_team_invite_notify(self, data: bytes) -> None:
        """处理收到邀请通知"""
        import team_pb2
        notify = team_pb2.TeamInviteNotify()
        notify.ParseFromString(data)
        invite = TeamInviteNotify(
            team_id=notify.team_id,
            inviter_id=notify.inviter_id,
            inviter_name=notify.inviter_name,
            inviter_level=notify.inviter_level,
        )
        self._pending_invites.append(invite)
        logger.info(f"Received invite from {notify.inviter_name} to team {notify.team_id}")
        if self.on_invite_received:
            self.on_invite_received(invite)

    def on_team_accept_resp(self, data: bytes) -> None:
        """处理接受邀请响应"""
        import team_pb2
        resp = team_pb2.TeamAcceptResp()
        resp.ParseFromString(data)
        if resp.code == 0:
            self._team_id = resp.members[0].team_id if resp.members else 0
            self._update_members(resp.members)
            # Remove pending invite
            self._pending_invites = [
                i for i in self._pending_invites if i.team_id != self._team_id
            ]
            logger.info(f"Joined team {self._team_id}")
            if self.on_member_update:
                self.on_member_update()

    def on_team_reject_resp(self, data: bytes) -> None:
        """处理拒绝邀请响应"""
        import team_pb2
        resp = team_pb2.TeamRejectResp()
        resp.ParseFromString(data)
        # Remove from pending
        # team_id not in resp, clean up based on accept req context

    def on_team_leave_resp(self, data: bytes) -> None:
        """处理离开队伍响应"""
        import team_pb2
        resp = team_pb2.TeamLeaveResp()
        resp.ParseFromString(data)
        if resp.code == 0:
            self._clear_team()
            logger.info("Left team")

    def on_team_kick_resp(self, data: bytes) -> None:
        """处理踢出响应"""
        import team_pb2
        resp = team_pb2.TeamKickResp()
        resp.ParseFromString(data)
        if resp.code != 0:
            logger.warning(f"Kick failed: code={resp.code}")
            if self.on_error:
                self.on_error(resp.code, "踢出失败")

    def on_team_info_resp(self, data: bytes) -> None:
        """处理队伍信息响应"""
        import team_pb2
        resp = team_pb2.TeamInfoResp()
        resp.ParseFromString(data)
        if resp.code == 0:
            self._team_id = resp.team_id
            self._leader_id = resp.leader_id
            self._status = resp.status
            self._cave_level = resp.cave_level
            self._update_members(resp.members)
            logger.info(f"Team info: id={self._team_id}, members={len(self._members)}")

    def on_team_member_update(self, data: bytes) -> None:
        """处理成员变更通知"""
        import team_pb2
        notify = team_pb2.TeamMemberUpdate()
        notify.ParseFromString(data)
        self._update_members(notify.all_members)
        logger.info(f"Member update: type={notify.update_type}")
        if self.on_member_update:
            self.on_member_update()

    def on_team_leader_change(self, data: bytes) -> None:
        """处理队长变更通知"""
        import team_pb2
        notify = team_pb2.TeamLeaderChange()
        notify.ParseFromString(data)
        self._leader_id = notify.new_leader_id
        # Update member is_leader flags
        for m in self._members:
            m.is_leader = (m.player_id == self._leader_id)
        logger.info(f"Leader changed to {notify.new_leader_id}")
        if self.on_leader_change:
            self.on_leader_change(notify.new_leader_id)

    def on_team_status_update(self, data: bytes) -> None:
        """处理队伍状态变更"""
        import team_pb2
        notify = team_pb2.TeamStatusUpdate()
        notify.ParseFromString(data)
        self._status = notify.status
        self._cave_level = notify.cave_level

    # --- Internal helpers ---

    def _update_members(self, proto_members) -> None:
        """从 proto 成员列表更新内存"""
        self._members = []
        for m in proto_members:
            member = TeamMember(
                player_id=m.player_id,
                role_name=m.role_name,
                level=m.level,
                online=m.online,
                is_leader=m.is_leader,
                current_hp=m.current_hp,
                max_hp=m.max_hp,
            )
            self._members.append(member)
        if self._members:
            self._leader_id = next((m.player_id for m in self._members if m.is_leader), 0)

    def _clear_team(self) -> None:
        """清除队伍状态"""
        self._team_id = 0
        self._leader_id = 0
        self._members.clear()
        self._status = 0
        self._cave_level = 0
